### PR TITLE
Adds support for storage parameter types

### PR DESCRIPTION
**Pull request by bkchr**
This pr adds a new parameter types type, the storage parameter types.
This parameter type supports loading the value from the storage or
returning the given default value.

### CODE REVIEW

**Comment by xlc** [comment]
Those are not exposed in metadata and SDK will have trouble inspect those values.

**Comment by bkchr** [comment]
> Those are not exposed in metadata and SDK will have trouble inspect those values.

I know. Not sure what "SDK" is here. The idea is more to use it in test runtimes to change behavior of them. However, if wanted we can probably expose it in the metadata later.

**Comment by xlc** [comment]
SDK I mean polkadot.js or alike libraries.

If this is a test only thing, then I don't see much reason to expose them in metadata.

**Comment by bkchr** [comment]
> If this is a test only thing, then I don't see much reason to expose them in metadata.

You could reason that some chains want to make these parameter types easily changeable by the council.

**Comment by thiolliere** [comment]
the doc of Get says:

> A trait for querying a single fixed value from a type.
> Return a constant value.

I guess this now should be undertood as constant inside the block ?

File: frame/support/src/lib.rs [hunk 1]
```diff
@@ -84,8 +84,21 @@ pub use sp_runtime::{self, ConsensusEngineId, print, traits::Printable};
...
+///
+/// - Using `const` to create a parameter type that provides a `const` getter.
+///   It is required that the `value` is const.
+///
+/// - Declare the parameter type without `const` to have more freedom when creating the value.
+///
+/// - Using `storage` to create a storage parameter type. This type is special as it tries to
+///   load the value from the storage under a fixed key. If the value could not be found in the
+///   storage, the given default value will be returned. It is required that the value implements
+///   [`Encode`](codec::Encode) and [`Decode`](codec::Decode).
```
Comment (kianenigma) [inline frame/support/src/lib.rs:99]: would be good to mention how the key is generated?

File: frame/support/src/lib.rs [hunk 2]
```diff
@@ -84,8 +84,24 @@ pub use sp_runtime::{self, ConsensusEngineId, print, traits::Printable};
...
+/// - Using `const` to create a parameter type that provides a `const` getter.
+///   It is required that the `value` is const.
+///
+/// - Declare the parameter type without `const` to have more freedom when creating the value.
+///
+/// - Using `storage` to create a storage parameter type. This type is special as it tries to
+///   load the value from the storage under a fixed key. If the value could not be found in the
+///   storage, the given default value will be returned. It is required that the value implements
+///   [`Encode`](codec::Encode) and [`Decode`](codec::Decode). The key for looking up the value
+///   in the storage is build using the following formular:
```
Comment (apopiak) [inline frame/support/src/lib.rs:100]:
```suggestion
///   in the storage is built using the following formular:
```

File: frame/support/src/lib.rs [hunk 3]
```diff
@@ -84,8 +84,24 @@ pub use sp_runtime::{self, ConsensusEngineId, print, traits::Printable};
...
+///
+/// - Declare the parameter type without `const` to have more freedom when creating the value.
+///
+/// - Using `storage` to create a storage parameter type. This type is special as it tries to
+///   load the value from the storage under a fixed key. If the value could not be found in the
+///   storage, the given default value will be returned. It is required that the value implements
+///   [`Encode`](codec::Encode) and [`Decode`](codec::Decode). The key for looking up the value
+///   in the storage is built using the following formular:
+///
+///   `twox_128(":" ++ NAME ++ ":")` where `NAME` is the name that is passed as type name.
```
Comment (thiolliere) [inline frame/support/src/lib.rs:102]: we have each module at `twox_128(modulename)` indeed having `:` prefixed and suffixed allow their name not to collide. looks fine but I wonder what is best.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**